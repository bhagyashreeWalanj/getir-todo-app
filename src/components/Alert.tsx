import React, { useEffect } from 'react'
import styles from '../styles/alert.module.scss'
import { default as classnames } from 'classnames'
import { CgDanger } from 'react-icons/cg'
import { TiTick } from 'react-icons/ti'

export interface IToDoAlert {
  type: string
  msg: string
  removeAlert: () => void
}

const Alert = ({ type, msg, removeAlert }: IToDoAlert) => {
  useEffect(() => {
    const timeout = setTimeout(() => {
      removeAlert()
    }, 3000)
    return () => clearTimeout(timeout)
  })

  return (
    <div
      className={classnames(
        type === 'success-edit'
          ? [styles.editSuccessAlert]
          : type === 'danger-edit'
          ? [styles.editErrorAlert]
          : [styles.alert],
      )}
    >
      <p
        className={classnames({
          [styles.errormsg]: type === 'danger',
          [styles.successmsg]: type === 'success',
          [styles.errormsgEdit]: type === 'danger-edit',
          [styles.successmsgEdit]: type === 'success-edit',
        })}
        id={`alertDiv`}
      >
        {type === 'danger' ? (
          <CgDanger className={styles.dangerIcon} />
        ) : (
          <TiTick />
        )}{' '}
        {msg}
      </p>
    </div>
  )
}

export default Alert
