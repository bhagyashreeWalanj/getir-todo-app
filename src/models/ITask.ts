export interface ITask {
    id: string;
    title: string;
    createdAt: string;
    completed: boolean;
  }
  