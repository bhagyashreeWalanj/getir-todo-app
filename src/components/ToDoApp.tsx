import React, { useState, useEffect } from 'react'
import styles from '../styles/toDoApp.module.scss'
import Alert from './Alert'
import { ITask } from '../models/ITask'

export interface AddToDoProps {
  tasks: ITask[]
  createToDo(title: string, tasks: ITask[]): void
  fetchTasks(): void
  clearTasks(): void
  error: string
}

const ToDoApp = ({
  tasks,
  createToDo,
  fetchTasks,
  clearTasks,
  error,
}: AddToDoProps) => {
  const [alert, setAlert] = useState({ show: false, msg: '', type: '' })
  const [title, setTitle] = useState('')

  const showAlert = (show = false, type = '', msg = '') => {
    setAlert({ show, type, msg })
  }

  // Method to add new task
  const handleCreateToDo = (title: string) => {
    if (title !== '') {
      createToDo(title, tasks)
      if (error !== null) {
        showAlert(true, 'danger', error)
        return
      } else {
        showAlert(true, 'success', 'item added to the list')
        setTitle('')
      }
    } else {
      showAlert(true, 'danger', 'Please add task to save')
      return
    }
  }

  // Method to empty task list
  const handleClearAllTasks = () => {
    clearTasks()
    showAlert(true, 'success', 'List Empty SuccessFully !!')
  }

  useEffect(() => {
    fetchTasks()
  }, [fetchTasks])

  return (
    <div>
      <div className={styles.container}>
        {alert.show && <Alert {...alert} removeAlert={showAlert} />}
        <div className={styles.content}>
          <header className={styles.header}>TO DO LIST</header>
          <div className={styles.inputDivContainer}>
            <input
              className={styles.inputContainer}
              placeholder="Title"
              value={title}
              id={'addToDoInput'}
              onChange={(event) => setTitle(event.target.value)}
            />

            <button
              id={'addButton'}
              type="submit"
              onClick={() => handleCreateToDo(title)}
              className={styles.addButton}
            >
              TO DO
            </button>
            <button
              type="submit"
              onClick={() => handleClearAllTasks()}
              className={styles.clearAllButton}
            >
              CLEAR ALL
            </button>
          </div>
        </div>
      </div>
    </div>
  )
}

export default ToDoApp
