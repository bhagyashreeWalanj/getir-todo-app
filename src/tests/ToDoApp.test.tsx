import React from 'react'
import { shallow } from 'enzyme'
import ToDoApp from '../components/ToDoApp'
import Alert from '../components/Alert'
import { ITask } from '../models/ITask'

describe('Testing ToDoApp Component', () => {
  const createToDoFn = jest.fn()
  const fetchTasksFn = jest.fn()
  const clearTasks = jest.fn()

  const taskList: ITask[] = [
    {
      id: '1',
      title: 'Hello World 1',
      createdAt: '01/05/2022',
      completed: true,
    },
    {
      id: '2',
      title: 'Hello World 2',
      createdAt: '01/05/2022',
      completed: false,
    },
  ]

  const wrapper = shallow(
    <ToDoApp
      tasks={taskList}
      error={''}
      createToDo={createToDoFn}
      fetchTasks={fetchTasksFn}
      clearTasks={clearTasks}
    />,
  )

  const container = wrapper.find('.inputContainer')

  it('should display to Add ToDo task Component', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should add new Task without Error', () => {
    container.simulate('change', { target: { value: 'sam' } })
    wrapper.find('.addButton').simulate('click')
    expect(createToDoFn).toBeCalledTimes(1)
  })

  it('Should clear the input', () => {
    const inputValue = container.props().value
    container.simulate('change', { target: { value: 'sam' } })
    wrapper.find('.addButton').simulate('click')
    expect(inputValue).toBe('')
  })

  it('should not call click buttton when value is null', () => {
    container.simulate('change', { target: { value: '' } })
    wrapper.find('.addButton').simulate('click')
    expect(createToDoFn).toBeCalledTimes(0)
  })

  it('should display error on empty field button click', () => {
    container.simulate('change', { target: { value: '' } })
    wrapper.find('.addButton').simulate('click')

    expect(wrapper.find(Alert)).toHaveLength(1)
    // display error
    expect(wrapper.find(Alert).dive().find('#alertDiv').text()).toContain(
      'Please add task to save',
    )
  })

  it('should clear All Tasks from the list', () => {
    wrapper.find('.clearAllButton').simulate('click')
    expect(clearTasks).toBeCalledTimes(1)
    expect(wrapper.find(Alert).dive().find('#alertDiv').text()).toContain(
      'SuccessFully',
    )
  })
})
