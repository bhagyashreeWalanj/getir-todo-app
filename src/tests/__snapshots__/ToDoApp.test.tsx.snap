// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ToDoApp Component should display to Add ToDo task Component 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="content"
    >
      <header
        className="header"
      >
        TO DO LIST
      </header>
      <div
        className="inputDivContainer"
      >
        <input
          className="inputContainer"
          id="addToDoInput"
          onChange={[Function]}
          placeholder="Title"
          value=""
        />
        <button
          className="addButton"
          id="addButton"
          onClick={[Function]}
          type="submit"
        >
          TO DO
        </button>
        <button
          className="clearAllButton"
          onClick={[Function]}
          type="submit"
        >
          CLEAR ALL
        </button>
      </div>
    </div>
  </div>
</div>
`;
